src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Dsa.cs

```diff
@@ -0,0 +1,159 @@
+        /// <summary>
+        /// Return the size of the 'r' or 's' signature fields in bytes.
+        /// </summary>
+        internal static int DsaSignatureFieldSize(SafeDsaHandle dsa)
+        {
+            int size = DsaSizeQ(dsa);
+            Debug.Assert(size * 2 < DsaEncodedSignatureSize(dsa));
+            return size;
+        }
+
+        [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_DsaSizeP")]
+        private static extern int DsaSizeP(SafeDsaHandle dsa);
+        /// <summary>
+        /// Return the size of the key in bytes.
+        /// </summary>
+        internal static int DsaKeySize(SafeDsaHandle dsa)
+        {
+            int keySize = DsaSizeP(dsa);
+
+            // Assume an even multiple of 8 bytes (OpenSsl makes this assumption as well) in case
+            // there were some leading zero bytes truncated due to BIGNUM semantics.
+            // This minimizes the change the key size is wrong due to leading zero bits; there is only
+            // a 2^64 chance that the key has 64 leading zero bits and thus the keysize will be wrong.
+            keySize = (keySize + 7) / 8 * 8 * 8;
+            return keySize;
+        }
+
+        [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_DsaSign")]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        internal static extern bool DsaSign(SafeDsaHandle dsa, byte[] hash, int hashLength, byte[] refSignature, out int outSignatureLength);
+
+        [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_DsaVerify")]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        internal static extern bool DsaVerify(SafeDsaHandle dsa, byte[] hash, int hashLength, byte[] signature, int signatureLength);
+
+        internal static DSAParameters ExportDsaParameters(SafeDsaHandle key, bool includePrivateParameters)
+        {
+            Debug.Assert(
+                key != null && !key.IsInvalid,
+                "Callers should check the key is invalid and throw an exception with a message");
+
+            if (key == null || key.IsInvalid)
+            {
+                throw new CryptographicException();
+            }
+
+            IntPtr p_bn, q_bn, g_bn, y_bn, x_bn; // these are not owned
+            int    p_cb, q_cb, g_cb, y_cb, x_cb;
+
+            bool refAdded = false;
+            try
+            {
+                key.DangerousAddRef(ref refAdded); // Protect access to the *_bn variables
+
+                if (!GetDsaParameters(key,
+                    out p_bn, out p_cb,
+                    out q_bn, out q_cb,
+                    out g_bn, out g_cb,
+                    out y_bn, out y_cb,
+                    out x_bn, out x_cb))
+                {
+                    throw new CryptographicException();
+                }
+
+                // Match Windows semantics where p, g and y have same length
+                int pgy_cb = GetMax(new[] { p_cb, g_cb, y_cb });
```

Comment:
It'd be nice if `GetMax` had some overloads for a common number of params (e.g. at least for 2 and 3) that could be called without having to allocate a temp array.